README.markdown
# `uhttpc` - µHTTP client library

`uhttpc` is a simple low-level and lightweight Haskell HTTP 1.1
library providing the bare minimum required for HTTP benchmarking
purposes.

This is **not** a RFC compliant HTTP client library and **shall not**
be used as a general purpose HTTP implementation!

# `uhttpc-bench`

This Cabal package comes with an executable `uhttpc-bench` which
represents an `ab`/`weighttpd`-style HTTP benchmarking tool:

    uhttpc-bench - a Haskell-based ab/weighttp-style webserver benchmarking tool

    Simple HTTP benchmark tool similiar to ab and weighttp

    uttpc-bench [OPTIONS] <url>

    Common flags:
      -n=num                     number of requests    (default: 1)
      -t=num                     threadcount           (default: 1)
      -c=num                     concurrent clients    (default: 1)
      -k                         enable keep alive
              --csv=FILE         dump request timings as CSV (RFC4180) file
              --user-agent=ITEM  specify User-Agent    (default: "httpc-bench")
      -H=str                     add header to request
      -v      --verbose          enable more verbose statistics and output
              --no-stats         disable statistics
      -p=FILE                    perform POST request with file-content as body
      -l=FILE                    perform a POST request per line, no quoting,
                                 round-robin, each client independently
      -?      --help             Display help message
      -V      --version          Print version information

# How to use it

First, install the `uhttpc` package
(This requires GHC 7.6.x or later)

    $ cabal install uhttpc

A simple example invocation:

    $ uhttpc-bench -n 100000 -t1 -c3 -k http://localhost/

    uhttpc-bench - a Haskell-based ab/weighttp-style webserver benchmarking tool

    starting benchmark...
    finished in 2.314289 seconds, 100000 reqs (3 conns), 43209.8 req/s received
    status codes: 100000 HTTP-200
    data received: 36205.098 KiB/s, 85800000 bytes total (24600000 bytes http + 61200000 bytes content)
    rtt min/avg/max = 0.034/0.068/4.252 ms

For comparison, [`weighttp`](https://github.com/lighttpd/weighttp) can be invoked with the very same arguments (in this case at least):

    $ weighttp  -n 100000 -t1 -c3 -k http://localhost/

    weighttp - a lightweight and simple webserver benchmarking tool

    starting benchmark...
    spawning thread #1: 3 concurrent requests, 100000 total requests

    finished in 2 sec, 333 millisec and 421 microsec, 42855 req/s, 35908 kbyte/s
    requests: 100000 total, 100000 started, 100000 done, 100000 succeeded, 0 failed, 0 errored
    status codes: 100000 2xx, 0 3xx, 0 4xx, 0 5xx
    traffic: 85800000 bytes total, 24600000 bytes http, 61200000 bytes data

Another example invocation of `uhttpc-bench`:

    $ uhttpc-bench http://www.google.com/ -v -c8 -n1000

    uhttpc-bench - a Haskell-based ab/weighttp-style webserver benchmarking tool

    using 88-byte request header (+ 0-byte body):
     "GET / HTTP/1.1\r\nHost: www.google.com:80\r\nUser-Agent: uhttpc-bench\r\nConnection: close\r\n\r\n"

    starting benchmark...

    per-client stats:

     client spawned +0.000008 s, 125 reqs (125 conns), 8.1 req/s, finished in 15.369619 s
     rtt min/avg/med/max = 99.586/122.902/122.109/167.193 ms

     client spawned +0.000258 s, 125 reqs (125 conns), 8.1 req/s, finished in 15.383278 s
     rtt min/avg/med/max = 100.494/123.012/121.648/184.555 ms

     client spawned +0.000292 s, 125 reqs (125 conns), 8.1 req/s, finished in 15.365843 s
     rtt min/avg/med/max = 98.721/122.872/121.668/163.810 ms

     client spawned +0.000327 s, 124 reqs (124 conns), 8.1 req/s, finished in 15.346788 s
     rtt min/avg/med/max = 103.008/123.709/122.834/174.003 ms

     client spawned +0.000366 s, 126 reqs (126 conns), 8.2 req/s, finished in 15.367205 s
     rtt min/avg/med/max = 95.255/121.907/120.736/152.943 ms

     client spawned +0.000403 s, 124 reqs (124 conns), 8.1 req/s, finished in 15.357957 s
     rtt min/avg/med/max = 97.730/123.800/123.569/162.326 ms

     client spawned +0.000434 s, 125 reqs (125 conns), 8.1 req/s, finished in 15.388717 s
     rtt min/avg/med/max = 103.056/123.055/121.961/162.419 ms

     client spawned +0.000461 s, 126 reqs (126 conns), 8.2 req/s, finished in 15.394365 s
     rtt min/avg/med/max = 102.114/122.123/121.289/151.520 ms

    finished in 15.394867 seconds, 1000 reqs (1000 conns), 65.0 req/s received
    status codes: 1000 HTTP-302
    data received: 63.498 KiB/s, 1001000 bytes total (783000 bytes http + 218000 bytes content)
    rtt 2/9|25/50/75|91/98-th quantile = 103.737/108.696 | 115.124/121.688/129.944 | 137.715/148.159 ms
    rtt min/avg/max = 95.255/122.919/184.555 ms


The `--csv` option allows to export the raw measurement data in format
suitable for offline analysis with statistical tools such as [`R`](http://www.r-project.org).


CHANGELOG.md
#### 0.1.1.1

- Fix incorrectly constructed `Host:` header in `uhttpc-bench` executable
- Optimize/improve implementation of `getPOSIXTimeSecs` & `getPOSIXTimeUSecs`
- Add support for `network-3.x`

### 0.1.1.0

- `uhttpc-bench`: Add support for round-robin POST requests from file
- `uhttpc-bench`: Converted to `optparse-applicative`
- Fix `ssUnread` being incorrect when push-back buffer is non-empty
- Minor optimization reducing `IORef` contention

## 0.1.0.0

- Initial release


src-exe/uhttpc-bench.hs
{-# LANGUAGE Arrows             #-}
{-# LANGUAGE BangPatterns       #-}
{-# LANGUAGE DeriveDataTypeable #-}
{-# LANGUAGE OverloadedStrings  #-}
{-# LANGUAGE RecordWildCards    #-}

module Main (main) where

import           Control.Concurrent
import           Control.Concurrent.Async
import           Control.DeepSeq            (deepseq)
import           Control.Monad
import           Data.ByteString            (ByteString)
import qualified Data.ByteString            as B
import qualified Data.ByteString.Char8      as B8
import           Data.Data
import           Data.IORef
import           Data.List
import           Data.Maybe
import           Data.Monoid                as Mon ((<>))
import           Data.Ord
import           Data.Word
import           GHC.Conc.Sync              (getNumProcessors, numCapabilities)
import           Network.HTTP.MicroClient
import           Network.Socket             (SockAddr)
import           Options.Applicative
import           Options.Applicative.Arrows
import           System.IO
import           System.Mem                 (performGC)
import           Text.Printf

-- |Timestamp in seconds since POSIX epoch
getTS :: IO TS
getTS = getPOSIXTimeSecs

type TS = Double

timeIO :: IO t -> IO (TS, TS, t)
timeIO act = do
    ts0 <- getTS
    res <- act
    tsd <- fmap (subtract ts0) getTS
    return $! tsd `seq` (ts0,tsd,res)

fI :: (Integral a, Num b) => a -> b
fI = fromIntegral

data Args = Args
    { argNumReq    :: Word
    , argThrCnt    :: Int
    , argClnCnt    :: Word
    , argKeepAlive :: Bool
    , argCsv       :: FilePath
    , argUA        :: String
    , argHdrs      :: [String]
    , argVerbose   :: Bool
    , argNoStats   :: Bool
    , argLAddr     :: String
    , argWait      :: Double
    , argUrl       :: String
    , argPostFn    :: FilePath
    , argPostList  :: FilePath
    } deriving (Show,Data,Typeable)

argsParser :: Parser Args
argsParser = runA $ proc () -> do
    argNumReq <- asA (option auto $ value 1 <> metavar "NUM" Mon.<> short 'n'
                      <> help "number of requests" <> showDefault) -< ()

    argThrCnt <- asA (option auto $ value numCapabilities <> metavar "NUM" <> short 't'
                      <> help "thread-count" <> showDefault) -< ()

    argClnCnt <- asA (option auto $ value 1 <> metavar "NUM" <> short 'c'
                      <> help "concurrent clients" <> showDefault) -< ()

    argKeepAlive <- asA (switch $ short 'k' <> help "enable keep alive") -< ()

    argHdrs <- asA (many (option str $ metavar "STR" <> short 'H'
                    <> help "add header to request")) -< ()

    argUA <- asA (option str $ value "httpc-bench" <> metavar "UA" <> long "user-agent"
                  <> help "specify User-Agent header" <> showDefault) -< ()

    argCsv <- asA (option str $ value "" <> metavar "FILE" <> long "csv"
                   <> help "dump request timings as CSV (RFC4180) file") -< ()

    argVerbose <- asA (switch $ short 'v' <> long "verbose"
                       <> help "enable more verbose statistics and output") -< ()

    argNoStats <- asA (switch $ long "no-stats"
                       <> help "disable statistics") -< ()

    argPostFn <- asA (option str $ value "" <> metavar "FILE" <> short 'p'
                      <> help "perform POST request with file-content as body") -< ()

    argPostList <- asA (option str $ value "" <> metavar "FILE" <> short 'l'
                        <> help "perform a POST request per line, no quoting, round-robin, each client independently") -< ()

    argLAddr <- asA (option str $ value "" <> metavar "IPADDR" <> long "local-addr"
                     <> help "set specific source address for TCP connections") -< ()

    argWait <- asA (option auto $ value 0.0 <> metavar "SECS" <> long "wait"
                    <> help "wait time between requests (per client) in seconds"
                    <> showDefault) -< ()

    argUrl <- asA (argument str (metavar "<URL>")) -< ()

    returnA -< Args {..}


main :: IO ()
main = runInUnboundThread $ do
    putStrLn "uhttpc-bench - a Haskell-based ab/weighttp-style webserver benchmarking tool\n"

    pargs <- execParser (info (helper <*> argsParser)
                              (fullDesc <> header "Simple HTTP benchmarking tool similiar to 'ab' and 'weighttp'")
                        )
    let verbose = argVerbose pargs

    (hostname,portnum,urlpath) <- either fail return $ splitUrl (argUrl pargs)

    -- sanity-check CLI arguments
    unless (argNumReq pargs > 0) $
        fail "request-count must be > 0"

    unless (argClnCnt pargs > 0) $
        fail "client-count must be > 0"

    unless (argThrCnt pargs > 0) $
        fail "thread-count must be > 0"

    unless (argNumReq pargs >= argClnCnt pargs) $
        fail "request-count must be >= client-count"

    -- setup socket address
    sa <- getSockAddr hostname portnum
    lsa <- if null (argLAddr pargs)
           then return Nothing
           else fmap Just $ getSockAddr (argLAddr pargs) 0

    when verbose $
        printf "connecting to %s (using local address %s)\n" (show sa) (maybe "*:*" show lsa)

    -- set up worker threads
    when (fI (argThrCnt pargs) /= numCapabilities) $
        setNumCapabilities (fI $ argThrCnt pargs)
    thrcnt' <- getNumCapabilities

    when (argThrCnt pargs > 1 && thrcnt' /= argThrCnt pargs) $
        fail "failed to set worker tread-count"

    numCpus <- getNumProcessors

    when (thrcnt' > numCpus) $
        printf "*WARNING* thread-count (%u) > cpu cores (%u)\n" thrcnt' numCpus

    when (thrcnt' > 1) $
        printf "using up to %u Haskell execution threads\n" thrcnt'

    unless (fI (argClnCnt pargs) >= thrcnt') $
        putStrLn "*WARNING* client-count should be >= thread-count"

    ----------------------------------------------------------------------------
    let postOrGet = if null (argPostFn pargs) && null (argPostList pargs)
                    then GET
                    else POST

    when (verbose && not (null (argPostList pargs))) $
      putStrLn "reading the input file and preparing requests...\n"

    bodyList <- if not (null $ argPostList pargs)  -- overrides argPostFn
                then fmap (take (fI $ argNumReq pargs)  -- will at most use this
                           . map Just
                           . B8.lines)
                          $ B.readFile (argPostList pargs)
                else if not (null $ argPostFn pargs)
                then fmap (map Just . (: [])) $ B.readFile (argPostFn pargs)
                else return [Nothing]

    let hdrs | null (argUA pargs) = map B8.pack (argHdrs pargs)
             | otherwise          = "User-Agent: " <> B8.pack (argUA pargs) :
                                    map B8.pack (argHdrs pargs)
        mkRawreq :: Maybe B8.ByteString -> IO ByteString
        mkRawreq rawreqb = do
          let rawreqblen = maybe 0 B.length rawreqb
              msg = mkHttp11Req postOrGet
                                (B8.pack urlpath)
                                (B8.pack hostname <> ":" <> B8.pack (show portnum))
                                (argKeepAlive pargs)
                                hdrs
                                rawreqb
          -- Printed when preparing the requests, not when sending.
          when verbose $
            printf "using %d-byte request message (%d-byte body):\n %s\n\n"
                   (B.length msg) rawreqblen (show msg)
          return $! msg

    rawreqList <- mapM mkRawreq bodyList
    let reqList = cycle rawreqList

    -- We make sure the list is fully evaluated _before_ we send anything.
    rawreqList `deepseq`
      when (verbose && not (null (argPostList pargs))) $
        putStrLn $ show (length rawreqList) ++ " request(s) prepared\n"

    putStrLn "starting benchmark..."

    reqCounter <- newIORef (fI (argNumReq pargs) :: Int)

    let clnCnt = fI $ argClnCnt pargs

    performGC
    ts0' <- getTS

    tss  <- if clnCnt == 1
            then mapM (\() -> timeIO (doReqsAll lsa sa reqCounter reqList (argWait pargs))) [()]
            else mapConcurrently (\() -> timeIO (doReqs lsa sa reqCounter reqList (argWait pargs)))
                                 (replicate clnCnt ())
    ts1' <- getTS
    performGC

    numReqs' <- readIORef reqCounter
    unless (numReqs' + fI (argClnCnt pargs) == 0) $
        fail $ "virtual clients exited prematurely! " ++ show numReqs'

    unless (argNoStats pargs) $ do
        when verbose $ do
            putStrLn "\nper-client stats:\n"

            forM_ tss $ \(ts0,tdelta,ents') -> do
                let n' = length ts'
                    ts' = map (\e -> entRespDone e - entConnect e) ents'
                    conncnt = length [ () | Entry {..} <- ents', entConnect /= entReqStart ]

                _ <- printf " client spawned +%.6f s, %d reqs (%d conns), %.1f req/s, finished in %.6f s\n"
                            (ts0-ts0') n' conncnt (fI n' / tdelta) tdelta

                unless (null ts') $
                    printf " rtt min/avg/med/max = %.3f/%.3f/%.3f/%.3f ms\n"
                           (1000*minimum ts') (1000*avg ts') (1000*med ts') (1000*maximum ts')
                putStrLn ""

        let allents = concatMap (\(_,_,ents) -> ents) tss
            (rlen,clen) = foldl' (\(!r,!c) e -> (r + entRespRLen e,c + entRespCLen e)) (0,0) allents
            conncnt = length [ () | Entry {..} <- allents, entConnect /= entReqStart ]
            allts = [ entRespDone - entConnect | Entry {..} <- allents ]
            statcodes = hist $ map entRespCode allents

        _ <- printf "finished in %.6f seconds, %d reqs (%d conns), %.1f req/s received\n"
                    (ts1'-ts0') (length allents) conncnt (fI (argNumReq pargs) / (ts1'-ts0'))

        putStrLn $ "status codes: " ++ intercalate ", " [ show y ++ " HTTP-" ++ show x | (x,y) <- statcodes ]

        _ <- printf "data received: %.3f KiB/s, %u bytes total (%u bytes http + %u bytes content)\n"
                    (fI rlen / (1024 * (ts1'-ts0'))) rlen (rlen-clen) clen

        when verbose $ do
            let [q2,q9,q25,q50,q75,q91,q98] = summary7 (map (*1000) allts)
            printf "rtt 2/9|25/50/75|91/98-th quantile = %.3f/%.3f | %.3f/%.3f/%.3f | %.3f/%.3f ms\n" q2 q9 q25 q50 q75 q91 q98

        printf "rtt min/avg/max = %.3f/%.3f/%.3f ms\n"
               (1000*minimum allts) (1000*avg allts) (1000*maximum allts)

    ----------------------------------------------------------------------------
    -- CSV dumping
    unless (null $ argCsv pargs) $ do
        putStrLn "dumping CSV..."
        let allents = sortBy (comparing entConnect) $ concatMap (\(_,_,ents) -> ents) tss
        withFile (argCsv pargs) WriteMode $ \h -> do
            _ <- hPrintf h "timestamp_s,connect_ms,req_send_ms,resp_wait_ms,resp_recv_ms,resp_status,resp_total_bytes,resp_content_bytes,conn_id\r\n"
            forM_ allents $ \Entry {..} ->
                hPrintf h "%.6f,%.4f,%.4f,%.4f,%.4f,%d,%u,%u,%d\r\n"
                    entConnect
                    (1000 * (entReqStart  - entConnect))
                    (1000 * (entReqDone   - entReqStart))
                    (1000 * (entRespStart - entReqDone))
                    (1000 * (entRespDone  - entRespStart))
                    entRespCode
                    entRespRLen
                    entRespCLen
                    entConnId

    ----------------------------------------------------------------------------

    return ()
  where
    avg :: [Double] -> Double
    avg xs = sn / fI ln
      where
        (sn,ln) = foldl' go (0,0::Int) xs
        go (!s,!l) x = (s+x,l+1)

    med :: [Double] -> Double
    med xs = sort xs !! (l `div` 2)
      where
        l = length xs

    summary7 :: [Double] -> [Double]
    summary7 xs = atIndices (sort xs) is
      where
        l = length xs
        is = [ ((l-1)*q) `div` 100 | q <- [2,9,25,50,75,91,98] ]

    atIndices :: [a] -> [Int] -> [a]
    atIndices _ []      = []
    atIndices [] _      = error "atIndices"
    atIndices xs (i:is) = head xs' : atIndices xs' (map (subtract i) is)
      where
        xs' | i >= 0 = drop i xs
            | otherwise = error "atIndices2"

    hist xs = [ (head g, length g) | g <- group (sort xs) ]

-- |repeat executing 'doReq' until req-counter goes below 0
doReqs :: Maybe SockAddr -> SockAddr -> IORef Int -> [ByteString] -> Double -> IO [Entry]
doReqs lsa sa cntref reqList wtime = go Nothing [] reqList
  where
    go _ _ [] = fail "internal error: request list ran out"
    go ss a (req : rest) = do
        notDone <- decReqCounter cntref
        if notDone
        then do
            (ss',r) <- doReq lsa sa ss req
            when (wtimeUsecs > 0) $ threadDelay wtimeUsecs
            go ss' (r:a) rest
        else do
            maybe (return ()) ssClose ss
            return a

    wtimeUsecs = floor $ 1000000 * wtime

    decReqCounter cnt = atomicModifyIORef' cnt (\n -> (n-1,n>0))

-- | Version of 'doReqs' that does all requests as indicated by req-counter at once
--
-- This updates the IORef only once and thus avoids additional overhead
doReqsAll :: Maybe SockAddr -> SockAddr -> IORef Int -> [ByteString] -> Double -> IO [Entry]
doReqsAll lsa sa cntref reqList wtime = do
    n <- atomicModifyIORef' cntref (\n -> (n-(max 0 (n+1)),(max 0 n)))
    go Nothing [] n reqList
  where
    go _ _ _ [] = fail "internal error: request list ran out"
    go ss a cnt (req : rest)
        | cnt > 0 = do
            (ss',r) <- doReq lsa sa ss req
            when (wtimeUsecs > 0) $ threadDelay wtimeUsecs
            go ss' (r:a) (cnt-1) rest
        | otherwise = do
            maybe (return ()) ssClose ss
            return a

    wtimeUsecs = floor $ 1000000 * wtime

-- |single measurement entry
data Entry = Entry
    { entConnId    :: {-# UNPACK #-} !Int -- each connections gets a new id
    , entRespCode  :: {-# UNPACK #-} !HttpCode
    , entRespRLen  :: {-# UNPACK #-} !Word64 -- total amount of data received
    , entRespCLen  :: {-# UNPACK #-} !Word64 -- payload content-length
    , entConnect   :: {-# UNPACK #-} !TS   -- ts at connection-startup
    , entReqStart  :: {-# UNPACK #-} !TS   -- ts before sending request
    , entReqDone   :: {-# UNPACK #-} !TS   -- ts after sending request
    , entRespStart :: {-# UNPACK #-} !TS   -- ts when response starts being received
    , entRespDone  :: {-# UNPACK #-} !TS   -- ts when full response has been received
    } deriving (Show)

doReq :: Maybe SockAddr -> SockAddr -> Maybe SockStream -> ByteString -> IO (Maybe SockStream,Entry)
doReq lsa sa mss rawreq = do
    (ss,mt0) <- case mss of
        Just ss' -> return (ss',Nothing)
        Nothing  -> do
            t0' <- getTS
            ss' <- ssConnect lsa sa
            return (ss',Just t0')

    t1  <- getTS -- before request is sent (and socket is already connected)
    ssWrite rawreq ss
    t1' <- getTS -- after request has been sent

    _ <- ssPeek ss
    t2  <- getTS -- after first packet of response has been received

    rcnt0 <- ssReadCnt ss

    -- read HTTP response
    HttpResponse {..} <- recvHttpResponse ss

    t2'  <- getTS -- after full response has been received
    rcnt1 <- ssReadCnt ss

    -- sanity check
    tmp <- ssPeekBuf ss
    unless (B.null tmp) $
        fail "trailing garbage detected in server response -- aborting"

    -- keep-alive handling
    ss' <- if respKeepalive
           then
               return (Just ss)
           else do
               ssClose ss
               return Nothing

    let ent = Entry (ssId ss) respCode (rcnt1-rcnt0) respContentLen (fromMaybe t1 mt0) t1 t1' t2 t2'

    return $! strictPair ss' ent
  where
    strictPair !a !b = (a,b)
